providers/cloudns/cloudnsProvider.go

```diff
@@ -0,0 +1,228 @@
+package cloudns
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/StackExchange/dnscontrol/models"
+	"github.com/StackExchange/dnscontrol/providers"
+	"github.com/StackExchange/dnscontrol/providers/diff"
+	"github.com/miekg/dns/dnsutil"
+	"github.com/pkg/errors"
+	"strconv"
+)
+
+/*
+
+CloDNS API DNS provider:
+
+Info required in `creds.json`:
+   - auth-id
+   - auth-password
+
+*/
+
+var defaultNameServerNames = []string{
+	"pns101.cloudns.net",
+	"pns102.cloudns.net",
+	"pns103.cloudns.net",
+	"pns104.cloudns.net",
+}
+
+// NewCloudns creates the provider.
+func NewCloudns(m map[string]string, metadata json.RawMessage) (providers.DNSServiceProvider, error) {
+	api := &CloudnsApi{}
+
+	api.creds.id, api.creds.password = m["auth-id"], m["auth-password"]
+	if api.creds.id == "" || api.creds.password == "" {
+		return nil, errors.Errorf("missing ClouDNS auth-id and auth-password")
+	}
+
+	// Get a domain to validate authentication
+	if err := api.fetchDomainList(); err != nil {
+		return nil, err
+	}
+
+	return api, nil
+}
+
+var features = providers.DocumentationNotes{
+	providers.DocDualHost:            providers.Cannot(),
+	providers.DocOfficiallySupported: providers.Cannot(),
+	providers.DocCreateDomains:       providers.Can(),
+	providers.CanUseAlias:            providers.Can(),
+	providers.CanUseSRV:              providers.Can(),
+	// ClouDNS support all below, but not implemented yet
+	providers.CanUsePTR:   providers.Unimplemented(),
+	providers.CanUseSSHFP: providers.Unimplemented(),
+	providers.CanUseCAA:   providers.Unimplemented(),
+	providers.CanUseTLSA:  providers.Unimplemented(),
+}
+
+func init() {
+	providers.RegisterDomainServiceProviderType("CLOUDNS", NewCloudns, features)
+}
+
+// GetNameservers returns the nameservers for a domain.
+func (api *CloudnsApi) GetNameservers(domain string) ([]*models.Nameserver, error) {
+	return models.StringsToNameservers(defaultNameServerNames), nil
```

Comment:
wow. This question just opened local "Pandora's box" :) The short answer - I think ClouDNS **does** support another's NS servers. 

But honest answer - I don't think I really understand "nameservers" part at all. Nor in  dnscontrol  nor in general.  I tried to examine the code of other providers - didn't help... Will appreciate if you can give a hand.

1) What is the semantic and purpose of GetNameservers? Do we need it in case we have empty zone and we want to add provider's NS servers? Any other cases ?
2) should GetNameservers _always_ return, at least, list of providers NS ?

If `providers.DocDualHost:            providers.Can()`, does it mean

3) basically, provider allow to add  nameservers from another provider ? Or something more complex ?
4)  run TestDualProviders for such providers is a wrong behavior ? And should have

```golang 
if !providers.ProviderHasCabability(*providerToRun, providers.DocDualHost) {
  t.Log("Skipping TestDualProviders Tests because provider does not support DualHost")
  return
}
```

5) looks like my implementation of GetNameservers is wrong but all tests are fine. How can we test this case ?

Reply:
Ah.. nameservers are one of the least understood parts of DNS!  There's no shame in being confused.

* NS() lets you add an NS record to a zone, just like A() adds an A record to the zone.
* The NAMESERVER() directive adds an NS record to the parent zone. (Maybe it should have been called PARENTNAMESERVER()?)

Since the parent zone could be completely unrelated to the current zone, changes made by NAMESERVER() have to be done by an API call to the registrar, who then figures out what to do.  For example, if I change the NAMESERVER() for `stackoverflow.com`, the API has to talk to whoever controls ".com".    It can't do that directly, but it can make an API call to the DNS Registrar for stackoverflow.com, who knows how to talk to the .com people.

That's why NAMESERVER() is coded separately.

The DualHost stuff is related but different. Basically some DNS providers want to control everything about the (parent) nameservers.  That's fine for most people as the provider is doing the right thing, or preventing people from making typos, etc.  However some people (like me!) want to have two different DNS providers, and that means some NAMESERVERS() pointing to (in my case) ROUTE53, and some NAMESERVERS() pointing to GoogleDNS.  The whole DualHost thing is to verify that the provider permits that.

Reply:
Curiouser and curiouser! as said Alice... :)
 
I am very grateful for your explanation. And I spent several hours today googling and discuss this subject. Well, you are right, it's tricky. So I still need some help. :)

Lets say I am working with domain mydomain.com 

1) I changed implementation of GetNameservers in  77d867b and now it uses  ClouDNS API for [Available name servers  ](https://asia.cloudns.net/wiki/article/47/)

Lets say, this api now always return 

- pns101.cloudns.net
- pns102.cloudns.net

**Qestion1**: I have mydomain.com  with zero records. I do **first** run of dnscontrol. What should GetNameservers return? 
* 2 records  (pns101, pns102)
* zero records 

I believe 2 records.

2) ClouDNS **allow** to create NS records like
`mydomain.com NS  ns1.example.com`

**Question2**: If I have such record, what should GetNameservers returns ?
*  2 records (pns101, pns102)
*  3 records (pns101, pns102 +  ns1.example.com )

Reply:
I'm not sure I understand the questions.

If dnsconfig.js has zero NAMESERVER() records for a domain, it will try to remove all the nameservers using the API.
If dnsconfig.js has 1 or more NAMESERVER() records for a domain, it will use the API to set those as the nameservers (unless, of course, they're already set).

On the other hand, if you don't want dnscontrol to change the nameservers at all (which a lot of people prefer) then use the special Registrar called "NONE". It makes no changes.

```
var REG_THIRDPARTY = NewRegistrar('ThirdParty', 'NONE')
D("mydomain.com", REG_THIRDPARTY, ...)
```

Reply:
thanks for support :) Based on all answers ( and my tests with second BIND  provider and registrar), I believe, my current implementation of GetNameservers is correct for ClouDNS.  

If you think it's not correct, could you help me and give example of dnsconfig.js where I can see problem.